File: SafeSpace.java
import java.io.*;
import java.util.Hashtable;

class SafeSpace implements ISpace, Serializable {
    //data
    public SpaceType st = ISpace.SpaceType.SAFE;
    public int id;
    public ISpace next_space = null;
    public ISpace next_home = null;
    private Pawn pawn1 = null;
    private Pawn pawn2 = null;

    //constructor
 SafeSpace(int id) {
        this.id = id;
    }

    //functions
    public Pawn add_pawn(Pawn p) {
        if (this.pawn1 == null) {
            this.pawn1 = p;
            return null;
        } else if (this.pawn1.color.equals(p.color)){
            this.pawn2=p;
            return null;
        } else {
            return null;
        }
    }

    public Pawn remove_pawn(Pawn p) {
        if (p.equals(this.pawn1)) {
            this.pawn1 = this.pawn2;
            this.pawn2 = null;
            return pawn1;
        } else if (p.equals(this.pawn2)) {
            return pawn2;
        }
        else{
            return null;
        }
    }

    public ISpace get_next_space() {
        return this.next_space;
    }

    public void set_next_space(ISpace s) {
        this.next_space = s;
    }

    public ISpace get_next_home() {
        return this.next_home;
    }

    public void set_next_home(ISpace s) {
        this.next_home = s;
    }

    public ISpace.SpaceType get_space_type() {
        return this.st;
    }

    public String get_color() {
        return "";
    }
    
    public int get_id() {
        return this.id;
    }

    public Hashtable<String, Pawn> get_pawns() {
        Hashtable<String, Pawn> pawns = new Hashtable<String, Pawn>();
        pawns.put("pawn1", this.pawn1);
        pawns.put("pawn2", this.pawn2);
        return pawns;
    }

    public boolean has_blockade() {
        return (this.pawn1 != null && this.pawn2 != null);
    } 
}

File: python/SafeSpace.py
"""Implements the SafeSpace class (extends Space)"""

#Global

#Local
from Pawn import Pawn
from Space import Space

class SafeSpace(Space):
    def __init__(self, s_id):
        """Takes a space id"""
        Space.__init__(self, s_id)
        self.pawn1 = None
        self.pawn2 = None
        self.next_home = None #Space
    
    def add_pawn(self, pawn): #Pawn
        if self.pawn1 == None:
            self.pawn1 = pawn
            return None
        elif self.pawn1.color == pawn.color:
            self.pawn2 = pawn
            return None
        else:
            removed_pawn = self.remove_pawn(self.pawn1)
            self.pawn1 = pawn
            return removed_pawn

    def remove_pawn(self, pawn): #Pawn
        if pawn == self.pawn1:
            self.pawn1 = self.pawn2
            self.pawn2 = None
            return pawn
        elif pawn == self.pawn2:
            self.pawn2 = None
            return pawn
        else:
            return None

    def get_pawns(self):  #dict of pawns
        return {"pawn1": self.pawn1, "pawn2": self.pawn2}
    
    def has_blockade(self):  #boolean
        return self.pawn1 != None and self.pawn2 != None
 
if __name__ == "__main__":
    print("Safe Space")
